pkg/kubelet/metrics/collectors/resource_metrics.go

```diff
@@ -150,6 +159,14 @@ func (rc *resourceMetricsCollector) collectNodeMemoryMetrics(ch chan<- metrics.M
 		metrics.NewLazyConstMetric(nodeMemoryUsageDesc, metrics.GaugeValue, float64(*s.Memory.WorkingSetBytes)))
 }
 
+func (rc *resourceMetricsCollector) collectContainerStartTime(ch chan<- metrics.Metric, pod summary.PodStats, s summary.ContainerStats) {
+	if s.StartTime.Unix() == 0 {
+		return
+	}
+
+	ch <- metrics.NewLazyConstMetric(containerStartTimeDesc, metrics.GaugeValue, float64(s.StartTime.UnixNano()), s.Name, pod.PodRef.Name, pod.PodRef.Namespace)
```

Comment:
I think we should either use same format as:
* process_start_time metric or
* [timestampedMetric](https://github.com/kubernetes/kubernetes/blob/14e4ed07cf81a8a273c63943c29f9f54761daebb/vendor/github.com/prometheus/client_golang/prometheus/metric.go#L157)

Can you compare those two formats so we can discuss it in this PR? Would be great to see some real examples

Reply:
hi, @serathius, 

- `process_start_time_seconds` of kubelet does not reported as `timestampedMetrics`,:

```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics |grep process_start_time
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.62250734682e+09
```

- The `statsSummary` reported by `stats.SummaryProvider` provide no sampling timestamp of `container_start_time` metrics, since `container_start_time` is independent of sampling timestamp, does `time.Now()` make scence?

here is the real example metrics that using `time.Now()` as the sampling timestamp of `container_start_time`:
```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics/resource |grep  container_start_time
# HELP container_start_time [ALPHA] Start time of container in UNIX time
# TYPE container_start_time gauge
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18 1622509662905
container_start_time{container="calico-node",namespace="kube-system",pod="calico-node-fgztf"} 1.6225073746780928e+18 1622509662905
container_start_time{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-10.235.30.78"} 1.622507351582303e+18 1622509662905
container_start_time{container="kube-controller-manager",namespace="kube-system",pod="kube-controller-manager-10.235.30.78"} 1.6225073518739638e+18 1622509662905
container_start_time{container="kube-proxy",namespace="kube-system",pod="kube-proxy-10.235.30.78"} 1.6225073517178916e+18 1622509662905
container_start_time{container="kube-scheduler",namespace="kube-system",pod="kube-scheduler-10.235.30.78"} 1.6225073515772767e+18 1622509662906
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-0"} 1.6225073717213302e+18 1622509662906
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-1"} 1.6225073684144394e+18 1622509662906
```

If we do not use `timestampedMetrics`, `container_start_time` metrics will be exposed like this:
```
[root@server-78 ~]# curl -s  http://127.0.0.1:8001/api/v1/nodes/10.235.30.78/proxy/metrics/resource |grep  container_start_time
# HELP container_start_time [ALPHA] Start time of container in UNIX time
# TYPE container_start_time gauge
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18
container_start_time{container="calico-node",namespace="kube-system",pod="calico-node-fgztf"} 1.6225073746780928e+18
container_start_time{container="kube-apiserver",namespace="kube-system",pod="kube-apiserver-10.235.30.78"} 1.622507351582303e+18
container_start_time{container="kube-controller-manager",namespace="kube-system",pod="kube-controller-manager-10.235.30.78"} 1.6225073518739638e+18
container_start_time{container="kube-proxy",namespace="kube-system",pod="kube-proxy-10.235.30.78"} 1.6225073517178916e+18
container_start_time{container="kube-scheduler",namespace="kube-system",pod="kube-scheduler-10.235.30.78"} 1.6225073515772767e+18
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-0"} 1.6225073717213302e+18
container_start_time{container="test-pvc-busybox",namespace="default",pod="a-test-pvc-sts-1"} 1.6225073684144394e+18
```

Reply:
Ok, let's use the same format as process_start_time so ` time since unix epoch in seconds`, but let's make sure to keep the milisecond resolution. This differs a big from the last result you have shown which is in nanoseconds.

Difference is in unit `seconds` vs `nanoseconds` and can be seen in exponential `e+09` vs `e+18`. Before passing time value let's make sure we divide it to 10^9 to change the unit.

Example in seconds
```
process_start_time_seconds 1.62250734682e+09
```

Example In nanoseconds:
```
container_start_time{container="calico-kube-controllers",namespace="kube-system",pod="calico-kube-controllers-fcb4577b8-vtmhn"} 1.6225073641000604e+18
```

Let's also update the description of the metric to reflect the unit.
```
# HELP container_start_time [ALPHA] Start time of container in UNIX time
```
To 
```
# HELP container_start_time [ALPHA] Start time of the container since unix epoch in seconds.
```

Reply:
Thanks for your advise, and now the metrics `container_start_time` exposed as the example blow. 
```
# HELP container_start_time [ALPHA] Start time of the container since unix epoch in seconds
# TYPE container_start_time gauge
container_start_time{container="container_a",namespace="namespace_a",pod="pod_a"} 1.624396248302e+09
container_start_time{container="container_a",namespace="namespace_b",pod="pod_b"} 1.624395678302e+09
container_start_time{container="container_b",namespace="namespace_a",pod="pod_a"} 1.624396158302e+09
```

There ara hesitation about the metrics name, should we add the unit `second` to the metrics name, that is `container_start_time_seconds`, so it is more prometheus style?

Reply:
I think so, it also matches `process_start_time_seconds`. Good catch.